### ISSUE TITLE
Implement `nanoid()`

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#18612]
follow up to https://github.com/prisma/prisma-engines/pull/3556

**Comment by andrictham** [comment]
Is this still planned?

**Comment by Piglow19** [comment]
Hello, can we have a update on this feature please ?

**Comment by janpio** [comment]
Yes, we will do this in the future - but I can not tell you when that will be as we currently do not have this work prioritized.

**Comment by ctsstc** [comment]
For anyone following the tangled web here's some useful links from other related issues:

- Main NanoID Repo: https://github.com/ai/nanoid/blob/main/README.md
    - Other Languages: https://github.com/ai/nanoid/blob/main/README.md#other-programming-languages
- SQL function: https://github.com/viascom/nanoid-postgres/
- PSQL Extension: https://github.com/spa5k/uids-postgres
- Utilizing a custom generator: https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#dbgenerated
	- https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#override-default-value-behavior-for-supported-types

**Comment by itzsaga** [comment]
My team verified that `id String @id @default(nanoid())` works. Looks like it was added in [4.9.0](https://github.com/prisma/prisma/releases/tag/4.9.0). Pretty sure this can be closed @janpio.

**Comment by itzsaga** [comment]
@ctsstc just double checked and nothing has been done to the DB (spun op a fresh PostgeSQL container and ran against it to be safe, checked functions and what not in case a migration did something). No type errors when defining it:
<img width="387" alt="Screenshot 2023-12-06 at 9 17 40 PM" src="https://github.com/prisma/prisma/assets/11750061/90f5bfd5-36c5-4e3d-9365-f3379d89d561">
If I change it something else random there are errors:
<img width="440" alt="Screenshot 2023-12-06 at 9 18 08 PM" src="https://github.com/prisma/prisma/assets/11750061/893c4ec0-43cd-4c70-891a-4a448f87f119">
So there's some level of type checking what goes in there. I wish I knew the Prisma code base better to dig in and see if I could verify myself. 🤷🏻‍♂️

**Comment by vancerbtw** [comment]
Any updates?

**Comment by rickyraz** [comment]
any update please

**Comment by LilaRest** [comment]
**Prisma now supports NanoIDs 🎉** well... since _**1 year and a half**_.

I just realized that support for NanoID has been introduced since [Prisma 4.9.0](https://github.com/prisma/prisma/releases/tag/4.9.0) (Janv. 2023), but none of the conversations and issues about it have been updated, nor the Prisma documentation.

--- 

The function type seems to be something like that, where `length` is the number of characters in the generated NanoID.

```ts
nanoid(length: Int?): String
```

Enjoy!

**Comment by bngrid** [comment]
> Hey everyone, Alberto from Prisma here.
> 
> As @LilaRest already [pointed out](https://github.com/prisma/prisma/issues/18612#issuecomment-2363075065), `nanoid` support has already been out there for more than 1 year, and is documented [here](https://www.prisma.io/docs/orm/reference/prisma-schema-reference#nanoid). We apologise for our oversight on notifying you about the progress on this issue.

I found that using @default(nanoid()) in `type` of schema.prisma will report an error: 
Error parsing attribute "@default": The function `nanoid()` cannot be used on fields of type `String`.
"prisma": "^6.0.1"
"@prisma/client": "^6.0.1"
![image](https://github.com/user-attachments/assets/eb811f35-edf9-4915-8c9e-b2cfa1d2997e)

**Comment by krishna-santosh** [comment]
@bngrid Maybe the error is due to the fact that you are using it in a type instead of a model.

**Comment by bngrid** [comment]
> @bngrid Maybe the error is due to the fact that you are using it in a type instead of a model. The `@default()` attribute is only supported in model definitions, not type definitions.

But using @default(uuid()) or @default(cuid()) doesn't seem to result in an error.

**Comment by krishna-santosh** [comment]
@bngrid you are right, using `@default(nanoid())` is resulting in an error.

I suspect that `@default()` values specified on a `type` are being ignored, and there's no validation happening in such cases and the field is being treated as a `String`. However, since `@default(nanoid())` is relatively new, it might not be ignored in the same way.

I tested this locally with MongoDB,  using `@default(uuid())` and `@default(cuid())` and it had no effect. I was able to insert any type of string without issue.

If you're looking to enforce stricter validation on the field's value, I recommend using Zod to validate the data before inserting it into the database.

I’m not sure why `@default()` values are allowed in types if they serve no purpose.
